Implements the Syncthing  data mover (#134)
backube/volsync: docs/usage/syncthing/index.rst

This can be used to create a hub-and-spoke model for replication, or any other kind of network.

When a ReplicationSource is created and configured to sync a PVC with other peers, all of the connected peers will maintain their
own Volume containing the synced data. To detect file changes, Syncthing employs two methods: a filesystem watcher, which notifies 
Syncthing of any changes to the local filesystem, and a full filesystem scan which occurs routinely at a specified interval (default is an hour).
Syncthing itself is an "always-on" synchronization system, and so created ReplicationSources will never finish their syncs unless deleted.

VolSync uses a custom-built Syncthing mover which disables the use of relay servers and global announce, and relying instead on 
being provided with the addresses of other Syncthing peers directly.

.. note::
    Syncthing is peer-to-peer technology which connects to other peers directly rather than going through intermediary servers.
    Because Syncthing lacks centralization, file conflicts are resolved by favoring the `most recent version <https://docs.syncthing.net/users/syncing.html#conflicting-changes>`_.

> Communicating With Syncthing

Reviewer: Perhaps move this to an advanced configuration section? I don't think users would normally do this, correct?

  metadata:
    # this should be in the format: volsync-<REPLICATION_SOURCE_NAME>
    name: volsync-my-replication-source
  type: Opaque
  # all of these fields must be provided
  data:
    # loaded by Syncthing
    httpsKeyPEM: <your base64 encoded HTTPS private key>
    # loaded into Syncthing and used by VolSync as a root CA
    httpsCertPEM: <your base64 encoded HTTPS certificate>
    # The API key used by VolSync to authenticate API requests
    apikey: <base64-encoded API Key>
    # These fields are solely for securing the Web UI from being accessed.
    username: <base64-encoded username>
    password: <base64-encoded password>

Once you have deployed this secret in your intended namespace, you will then need to create the ReplicationSource
using the name you specified in the above Secret. For example, a custom secret named "volsync-my-replication-source"
would require you to name the ReplicationSource "my-replication-source".

.. note::
  This Secret must be created **before** creating the ReplicationSource.
  Otherwise, Syncthing will generate its own set of credentials and ignore yours.  

Configuring a ReplicationSource
===============================

> Here's an example of Syncthing-based ReplicationSource.

Reviewer: ```suggestion
Here's an example of a Syncthing-based ReplicationSource.
```

    :caption: ReplicationSource object configuring the peers it should connect to.

    ---
    apiVersion: volsync.backube/v1alpha1
    kind: ReplicationSource
    metadata:
      name: sync-todo-database
    spec:
      sourcePVC: todo-database
      syncthing:
        serviceType: ClusterIP
        # List of peers that this ReplicationSource should connect to.
        peers:
          # The Syncthing ID of the peer.
        - ID: GVONGZX-6FVQPEY-4QWTVLK-TXNJUHA-5UGA625-UBC7HZQ-P5BG2XJ-EHJ4XQ3
          # The address of the peer - this will be used as a data connection.
          address: tcp://10.96.55.168:22000
          # Whether or not the peer should introduce this ReplicationSource to other peers.
          introducer: false

The above ReplicationSource tells VolSync that it should use the Syncthing replication method in order
to sync the ``todo-database`` volume. 

This ReplicationSource tells VolSync to use a ``ClusterIP`` service to expose the Syncthing data port.
In order for Syncthing to connect to other instances outside of the cluster, you will need to either use
``serviceType: LoadBalancer``, or a submariner-type cross-cluster networking configuration.

> The above ReplicationSource specifies a single peer to sync its data with, however you can specify as many or as little peers as you'd like.

Reviewer: ```suggestion
The above ReplicationSource specifies a single peer to sync its data with, however you can specify as many or as few peers as you'd like.
```

To create a simple ReplicationSource without connecting to other peers, simply omit the ``peers`` field.

In order for two Syncthing-based ReplicationSources to connect to each other, each one must specify the other one in their ``peers`` list.

.. note::

>   Syncthing unions the data in the provided ``PersistentVolume`` with the data from other peers.
>   When two pieces of data have the same name, the most recent data will be favored.

Reviewer: ```suggestion
  Syncthing combines the set of files in the provided ``PersistentVolume`` with those from the other peers.
  When two files have the same name, the file with the most recent data will be favored.
```

Syncthing options
-----------------

Here are all of the options that can be specified for the Syncthing mover:

peers
   A list of the Syncthing devices this ReplicationSource should sync the ``sourcePVC`` with. The peers
   being listed must also specify this ReplicationSource's Syncthing details in their spec for them to
   successfully connect with one another. Each peer contains the following fields:

>    - ``ID`` - The peer's device ID, this is a base32 encoding of the peer's TLS certificate.

Reviewer: ```suggestion
   - ``ID`` - The peer's device ID.
```
